Changes Css.register API to accept string param

DubeySandeep opened the pull request:
<!--
  - Thanks for submitting code to Blockly!  Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## The basics

<!-- TODO: Verify the following, checking each box with an 'x' between the brackets: [x] -->

- [ ] I branched from develop
   - Branched from goog_module as the changes are related to migration and can create conflicts on develop branch ([ref](https://github.com/google/blockly/issues/5205#issuecomment-918412337)).
- [ ] My pull request is against develop
  - The pull-request is against goog_module as major migration changes are happening on goog_module branch ([ref](https://github.com/google/blockly/issues/5205#issuecomment-918412337)).
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

## The details
### Resolves

<!-- TODO: What Github issue does this resolve? Please include a link. -->
Fixes #5205 

### Proposed Changes

<!-- TODO: Describe what this Pull Request does.  Include screenshots if applicable. -->
This pull request changes Css.registry API to accept string params

#### Behavior Before Change

_This PR doesn't change any user-facing behavior._
#### Behavior After Change

<!--TODO: Image, gif or explanation of behavior after this pull request. -->

_This PR doesn't change any user-facing behavior._

### Reason for Changes

The existing Blockly.Css.registry function have following design issues:
    1. Accepts css styles as array of strings, which internally gets concat to create a single multiline string.
    https://github.com/google/blockly/blob/57ac0ae33bbd80da313f512c510cb9830ba0af7b/core/css.js#L28-L41
    2. Eslint-bypass flags are used to have better presentation of css-style in string-of-array structure.
    https://github.com/google/blockly/blob/f94b1db85018b5a7261fa01f7d04cd5c8367fe2c/core/zoom_controls.js#L517-L521

These design choices were made in the first place because there wasn't any simple way to use multiline strings in es5. Considering the codebase is migrating to es6 we are fixing the design to accept a string param for the Css.registry method.  
### Test Coverage

<!-- TODO: Please show how you have added tests to cover your changes,
  -        or tell us how you tested it. For each systems you tested,
  -        uncomment the systems in the list below.
  -->

Unable to test the changes manually as running `npm run build` is not updating the blockly_compressed.js file.

On running `npm test`, all tests passed except the mocha test which is failing on the goog_module branch itself
![image](https://user-images.githubusercontent.com/16653571/133504563-f9e44273-0102-4d2a-9c6f-4de81f190be5.png)
 
<!-- Tested on: -->
<!-- * Desktop Chrome -->
<!-- * Desktop Firefox -->
<!-- * Desktop Safari -->
<!-- * Desktop Opera -->
<!-- * Windows Internet Explorer 10 -->
<!-- * Windows Internet Explorer 11 -->
<!-- * Windows Edge -->

<!--
* Smartphone/Tablet/Chromebook (please complete the following information):
  * Device: [e.g. iPhone6]
  * OS: [e.g. iOS8.1]
  * Browser [e.g. stock browser, safari]
  * Version [e.g. 22]
-->

### Documentation

<!-- TODO: Does any documentation need to be created or updated because of this PR?
  -        If so please explain.
  -->

As this PR changes Css.registry API, there needs to be an update in the following ref docs once the good_module branch is merged to develop/master: 

https://developers.google.com/blockly/reference/js/Blockly.Css?hl=en#.CONTENT
https://developers.google.com/blockly/reference/js/Blockly.Css?hl=en#.register
https://developers.google.com/blockly/reference/js?hl=en#Blockly.Css

### Additional Information

<!-- Anything else we should know? -->

DubeySandeep commented on core/comment.js:
```diff
@@ -398,12 +398,17 @@ Comment.prototype.dispose = function() {
+    padding: 3px;
+    resize: none;
+    text-overflow: hidden;
+  }
+`);
```
```suggestion
  }`);
```
Should we follow this^ style?

DubeySandeep replied:
Leaving it as is because the existing structure shows a clear css syle till a line. Thanks for the info! :)

DubeySandeep commented on core/css.js:
```diff
@@ -57,502 +55,503 @@ const inject = function(hasCss, pathToMedia) {
   const mediaPath = pathToMedia.replace(/[\\/]$/, '');
-  text = text.replace(/<<<PATH>>>/g, mediaPath);
+  let cssContent = CONTENT.replace(/<<<PATH>>>/g, mediaPath);
+  // Cleanup the collected css content after injecting it to the DOM.
+  CONTENT = '';
```
Do we need this step? (not sure!)

alschmiedt replied:
I'm double checking with someone, but in the mean time I think it is fine to leave it.

NeilFraser replied:
Yup, that cleanup is great!

DubeySandeep commented on core/css.js:
```diff
@@ -57,502 +55,503 @@ const inject = function(hasCss, pathToMedia) {
-  top: 0;
-  left: 0;
-}`,
+let CONTENT = (`
```
Not sure whether this still needs to be in CAPS?

DubeySandeep replied:
Done!

DubeySandeep commented on core/css.js:
```diff
@@ -57,502 +55,503 @@ const inject = function(hasCss, pathToMedia) {
 
   /* Added as a separate rule with multiple classes to make it more specific
-     than a bootstrap rule that selects svg:root. See issue #1275 for context.
+      than a bootstrap rule that selects svg:root. See issue #1275 for context.
```
Undo this change!

DubeySandeep replied:
Fixed!

DubeySandeep commented on core/css.js:
```diff
@@ -57,502 +55,503 @@ const inject = function(hasCss, pathToMedia) {
-  `.blocklyText text {
-  cursor: default;
-}`,
+      ahead of block during a drag. This way the cursor is still a closed hand.
```
Ditto*

DubeySandeep replied:
Fixed!

DubeySandeep commented on core/css.js:
```diff
@@ -57,502 +55,503 @@ const inject = function(hasCss, pathToMedia) {
-.blocklyScrollbarHandle:hover {
-  fill: #bbb;
-}`,
+      certain length. This affects our sizing calculations of the text input.
+      Hiding the close icon to avoid that. */
```
Ditto*

DubeySandeep replied:
Fixed!

DubeySandeep commented on core/css.js:
```diff
@@ -57,502 +55,503 @@ const inject = function(hasCss, pathToMedia) {
+    padding-left: 28px;
+  }
+
+    /* BiDi override for the resting state. */
```
De-indent by 2

DubeySandeep replied:
Done!

DubeySandeep commented on core/toolbox/category.js:
```diff
@@ -13,7 +13,6 @@
 goog.module('Blockly.ToolboxCategory');
 goog.module.declareLegacyNamespace();
 
-const Blockly = goog.require('Blockly');
```
Removed this as eslint is reporting warning on goog_module branch

DubeySandeep commented on tests/deps.js:
```diff
@@ -66,7 +66,7 @@ goog.addDependency('../../core/events/events_viewport.js', ['Blockly.Events.View
 goog.addDependency('../../core/extensions.js', ['Blockly.Extensions'], ['Blockly.utils'], {'lang': 'es6', 'module': 'goog'});
 goog.addDependency('../../core/field.js', ['Blockly.Field'], ['Blockly.DropDownDiv', 'Blockly.Events', 'Blockly.Events.BlockChange', 'Blockly.Gesture', 'Blockly.MarkerManager', 'Blockly.Tooltip', 'Blockly.WidgetDiv', 'Blockly.browserEvents', 'Blockly.utils', 'Blockly.utils.Rect', 'Blockly.utils.Size', 'Blockly.utils.Svg', 'Blockly.utils.dom', 'Blockly.utils.style', 'Blockly.utils.userAgent'], {'l ...
-goog.addDependency('../../core/field_angle.js', ['Blockly.FieldAngle'], ['Blockly.Css', 'Blockly.DropDownDiv', 'Blockly.FieldTextInput', 'Blockly.WidgetDiv', 'Blockly.browserEvents', 'Blockly.fieldRegistry', 'Blockly.utils.KeyCodes', 'Blockly.utils.Svg', 'Blockly.utils.dom', 'Blockly.utils.math', 'Blockly.utils.object', 'Blockly.utils.userAgent']);
+goog.addDependency('../../core/field_angle.js', ['Blockly.FieldAngle'], ['Blockly.Css', 'Blockly.DropDownDiv', 'Blockly.FieldTextInput', 'Blockly.WidgetDiv', 'Blockly.browserEvents', 'Blockly.fieldRegistry', 'Blockly.utils.KeyCodes', 'Blockly.utils.Svg', 'Blockly.utils.dom', 'Blockly.utils.math', 'Blockly.utils.object', 'Blockly.utils.userAgent'], {'lang': 'es6'});
```
Autogenerated change, is it an expected change?

alschmiedt reviewed (approved):
Overall LGTM! I think one nit to fix and then it should be good to go.

Thanks for all the notes in the PR description.

DubeySandeep commented on core/toolbox/category.js:
```diff
@@ -642,79 +641,79 @@ ToolboxCategory.prototype.dispose = function() {
+
+  .blocklyToolboxDiv[dir="RTL"] .blocklyTreeRow {
+    margin-left: 8px;
+    padding-right: 0
```
Err... I missed it. Thanks a lot for catching it!

Done!

DubeySandeep commented:
@alschmiedt Thanks for the review, I've addressed your review comments ([commit](https://github.com/google/blockly/pull/5472/commits/ada4403e3972a733100239a102d8b501ebed063a)).

alschmiedt commented on core/css.js:
```diff
@@ -28,16 +28,14 @@ let injected = false;
+ * @param {string} cssContent Multiline CSS string.
  */
-const register = function(cssArray) {
+const register = function(cssContent) {
```
Can you actually update this function to take in either a `string` or an `Array`? If an Array is passed in we can just convert it to a string and then use the [warn](https://github.com/google/blockly/blob/7b3a3aa74379465690dfed9c6c9f766809cf95c6/core/utils/deprecation.js#L31) method to tell the developer the API will change soon. The deletion date should be September 2022.

gonfunko replied:
I took care of this.

alschmiedt reviewed (commented):
Sorry for the last minute change, but once this is updated I'll merge it in. Thanks!

alschmiedt commented:
Hi @DubeySandeep,

I was wondering if you had time in the next couple of days to finish up the last comment on this? 

Thanks!
